types/validator_set.go

```diff
@@ -370,52 +371,87 @@ func processChanges(origChanges []*Validator) (updates, removals []*Validator, e
+//   Note that this will be < 2 * MaxTotalVotingPower in case high power validators are removed and
+//   validators are added/ updated with high power values.
+//
+// numNewValidators -  number of new validators
 // err - non-nil if the maximum allowed total voting power would be exceeded
 //
-// 'updates' should be a list of proper validator changes, i.e. they have been verified
-// by processChanges for duplicates and invalid values.
-// No changes are made to the validator set 'vals'.
 func verifyUpdates(
 	updates []*Validator,
 	vals *ValidatorSet,
-) (updatedTotalVotingPower int64, numNewValidators int, err error) {
+	removedPower int64,
+) (tvpAfterUpdatesBeforeRemovals int64, numNewValidators int, err error) {
 
-	updatedTotalVotingPower = vals.TotalVotingPower()
+	// powerChanges is a temporary slice that eventually stores the changes, starting with the
+	// decreases in validator powers.
+	powerChanges := make([]powerChange, 0, len(updates))
+	powerIncreases := make([]powerChange, 0, len(updates))
 
 	for _, valUpdate := range updates {
 		address := valUpdate.Address
 		_, val := vals.GetByAddress(address)
 		if val == nil {
-			// New validator, add its voting power the the total.
-			updatedTotalVotingPower += valUpdate.VotingPower
+			// new validator
+			powerIncreases = append(powerIncreases, powerChange{delta: valUpdate.VotingPower, address: address})
 			numNewValidators++
 		} else {
-			// Updated validator, add the difference in power to the total.
-			updatedTotalVotingPower += valUpdate.VotingPower - val.VotingPower
+			// updated validator, compute the power change
+			delta := valUpdate.VotingPower - val.VotingPower
+			change := powerChange{delta: delta, address: address}
+			if delta < 0 {
+				powerChanges = append(powerChanges, change)
+			} else {
+				powerIncreases = append(powerIncreases, change)
+			}
 		}
 	}
 
-	overflow := updatedTotalVotingPower > MaxTotalVotingPower
-	if overflow {
-		err = fmt.Errorf(
-			"failed to add/update validator, total voting power would exceed the max allowed %v",
-			MaxTotalVotingPower)
-		return 0, 0, err
-	}
+	// powerChanges only includes power decreases, append the increases
+	powerChanges = append(powerChanges, powerIncreases...)
+
+	// compute the total voting power after removals
+	tvpAfterRemovals := vals.TotalVotingPower() - removedPower
 
-	return updatedTotalVotingPower, numNewValidators, nil
+	// compute the total voting power after updates starting with the power decreases
+	for _, powerChange := range powerChanges {
+		tvpAfterRemovals += powerChange.delta
+		if tvpAfterRemovals > MaxTotalVotingPower {
+			err = fmt.Errorf(
+				"failed to add/update validator %v, total voting power would exceed the max allowed %v",
+				powerChange.address, MaxTotalVotingPower)
+			return 0, 0, err
+		}
+	}
+	tvpAfterUpdatesBeforeRemovals = tvpAfterRemovals + removedPower
```

Comment:
shouldn't we subtract here? `before = after - removedDelta` makes more sense to me rather than `before = after + removedDelta`

Reply:
We want the power before removals so we add the power that is removed. removed was already substracted from "after".